### PR TITLE
Client does not hang

### PR DESCRIPTION
**Pull request by marcorossi5**
This PR adds a `wait_for_results` boolean flag to `Client.run_circuit` method to let the user choose if the client should hang or not waiting for the server returning the results.

### CODE REVIEW

File: src/qibo_client/qibo_client.py
Comment (scarrazza) [inline src/qibo_client/qibo_client.py:151]:
```suggestion
        wait_for_results: bool = False,
```

**Comment by scarrazza** [comment]
This is missing a function which downloads the results from the server:
```python
r = client.get_results(pid=''')
```

File: src/qibo_client/qibo_client.py
Comment (scarrazza) [inline src/qibo_client/qibo_client.py:151]:
```suggestion
        wait_for_results: bool = False,
```

**Comment by marcorossi5** [comment]
This new commit implements the Qiskit-like interface of QiboClient and tests.

Address #22, #41, #46, #49

**Comment by scarrazza** [comment]
@marcorossi5 could you please include the following changes too:
- update the example in https://github.com/qiboteam/qibo-client/blob/feature/client-does-not-hang/examples/run_error_job.py
- update the README file with the new workflow (job, .result())

**Comment by marcorossi5** [comment]
@scarrazza, coverage is 100% now. Can you merge?

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [inline src/qibo_client/qibo_job.py:44]: This message should appear only if the job is not done. Could you please add a check?

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [inline src/qibo_client/qibo_job.py:113]: May I ask you to add a __call__ method, because otherwise we have to do job.result().result in order to get the result content.

**Review by scarrazza — CHANGES_REQUESTED** [review]

File: src/qibo_client/qibo_job.py
Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:113]: The following is the IBMQ way:

```python
job = backend.run(...)
 
try:
    job_result = job.result()  # It will block until the job finishes.
    print("The job finished with result {}".format(job_result))
except JobError as ex:
    print("Something wrong happened!: {}".format(ex))
```

Do you want to keep it and add `result = job()` syntax as well?

File: src/qibo_client/qibo_job.py
Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:44]: If we want to log this message only once, I am forced to add another api request outside the while loop to ask for the job status. The code will be really ugly in my opinion. Do you have an alternative implementation?

Btw in case the job is done the the messages displayed by the function will immediately be the following:

````
Please wait until your job is completed...
Job COMPLETED
```

Which is fine as well in my opinion

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [reply src/qibo_client/qibo_job.py:113]: The thing that seems annoying to me is:
```
r = job.result()
state = r.result.state()
```
Does IBMQ does the same?

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:113]: You should do this:

```
job.state()
r = job.result()
```

`r` is a  `QiboJobResult` instance and does not have a `state()` method:

https://github.com/qiboteam/qibo-client/blob/feature/client-does-not-hang/src/qibo_client/qibo_job.py#L105-L112

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:113]: The `QiboJobResult.success` boolean value contains whether the job was successful or not. I don't know if it's an important information that much

**Comment by scarrazza** [comment]
@marcorossi5, do we have a `client.get_job(jobid)` or similar function?

**Comment by marcorossi5** [comment]
Not yet. It can be useful. Let me implement it in a separate branch after we merge this two if you agree

**Comment by scarrazza** [comment]
ok, thanks.

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [reply src/qibo_client/qibo_job.py:44]: @marcorossi5 if you can fix this, I will merge.

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:44]: If you want to log it once the price is the following code. Which makes a repeated API call outside the while loop. Are you ok with this? Personally I don't like it

```python
def wait_for_response_to_get_request(
    url: str, seconds_between_checks: T.Optional[int] = None, verbose: bool = False
) -> T.Tuple[requests.Response, QiboJobStatus]:
    if seconds_between_checks is None:
        seconds_between_checks = constants.SECONDS_BETWEEN_CHECKS
    response = QiboApiRequest.get(url, timeout=constants.TIMEOUT)
        job_status = convert_str_to_job_status(response.headers["Job-Status"])
    if not verbose and job_status not in in [QiboJobStatus.DONE, QiboJobStatus.ERROR]:
        logger.info("Please wait until your job is completed...")

    while True:
        response = QiboApiRequest.get(url, timeout=constants.TIMEOUT)
        job_status = convert_str_to_job_status(response.headers["Job-Status"])
        if verbose and job_status == QiboJobStatus.QUEUED:
            logger.info("Job QUEUING")
        if verbose and job_status == QiboJobStatus.RUNNING:
            logger.info("Job RUNNING")
        if job_status in [QiboJobStatus.DONE, QiboJobStatus.ERROR]:
            if verbose:
                logger.info("Job COMPLETED")
            return response, job_status
        time.sleep(seconds_between_checks)
```

Comment (scarrazza) [reply src/qibo_client/qibo_job.py:44]: Ok, any other alternative?

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:44]: An exotic alternative that defines an adaptor for the logger that allows the following syntax `logger.info("message", only_once=True)`. But it seems an overkill and probably not that maintenable.

Otherwise, leave it as it is is not an option?
In the case you call the function with a completed job you'll get the following response.

```
Please wait until your job is completed...
Job COMPLETED
```

Which is not that bad I guess. What is the downside of this?

Comment (scarrazza) [reply src/qibo_client/qibo_job.py:44]: Ok, please go ahead with the second option.

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:44]: You mean this? `logger.info("message", only_once=True)`

Comment (scarrazza) [reply src/qibo_client/qibo_job.py:44]: No, I mean adding the "Job COMPLETED" tag after the please wait, so the user gets more information.

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [reply src/qibo_client/qibo_job.py:113]: Yes, indeed, IBMQ is not asking the user to call results twice, e.g.:
```
result = job.result()
print(result.state())
```
while in this PR, we have to do:
```
result = job.result()
print(result.result.state())
```
which is annoying, no?

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:113]: Ok, what is the object type you want me to return with `result.state()`?

Comment (scarrazza) [reply src/qibo_client/qibo_job.py:113]: I think we should remap job.result() into Qibo's corresponding object, so instead of returning a `QiboJobResults`, return the content of the `result` attribute directly.

Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:113]: We can drop the `QiboJobResult` class and return  the `qibo.result.QuantumState` object if you prefer

File: src/qibo_client/qibo_job.py
Comment (marcorossi5) [reply src/qibo_client/qibo_job.py:44]: we can then remove the `if verbose` line just before `logger.info("Job COMPLETED")` so that this information is always printed even if the user didn't ask

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [reply src/qibo_client/qibo_job.py:113]: I don't this will work because, the return object can also be a measurement, so you should return wathever qibo creates as output. You could drop this object and just keep the result quantity if this class is not used anywhere else.

File: src/qibo_client/qibo_job.py
Comment (scarrazza) [inline src/qibo_client/qibo_job.py:28]: Do you really need this function outside the `QiboJob` class? If you move there as a private function `_wait_for_response...` you can access the status and avoid all the problems raised in the discussion above.

**Pull request merged**
